controllers/osbuildconfig_controller.go

```diff
@@ -70,15 +83,152 @@ func (r *OSBuildConfigReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		return ctrl.Result{}, nil
 	}
 
-	err = r.OSBuildCRCreator.Create(ctx, osBuildConfig)
+	newOSBuildInstanceIsNeeded, err := r.checkIfNewOSBuildInstanceIsNeeded(ctx, osBuildConfig, logger)
 	if err != nil {
-		if errors.IsNotFound(err) {
-			return ctrl.Result{RequeueAfter: 30 * time.Second}, err
+		return ctrl.Result{Requeue: true, RequeueAfter: RequeueForShortDuration}, nil
+	}
+
+	if newOSBuildInstanceIsNeeded {
+		// create new OSBuild instance
+		targetBuild := osBuildConfig.Spec.Details.TargetImage.TargetImageType
+		if targetBuild == osbuilderv1alpha1.EdgeInstallerImageType {
+			// start with edge-container OSBuild instance
+			targetBuild = osbuilderv1alpha1.EdgeContainerImageType
+		}
+		return r.createOSBuildInstance(ctx, logger, osBuildConfig, targetBuild)
+	} else {
+		return r.updateOSBuildConfigCurrentStatus(ctx, logger, osBuildConfig)
+	}
+}
+
+func (r *OSBuildConfigReconciler) checkIfNewOSBuildInstanceIsNeeded(ctx context.Context, osBuildConfig *osbuilderv1alpha1.OSBuildConfig, logger logr.Logger) (bool, error) {
+	userConfiguration := r.getSortedUserConfiguration(osBuildConfig)
+	isNewOSBuildInstanceNeeded := false
+	patch := client.MergeFrom(osBuildConfig.DeepCopy())
+
+	if osBuildConfig.Status.LastKnownUserConfiguration == nil || !cmp.Equal(osBuildConfig.Status.LastKnownUserConfiguration, &userConfiguration) {
+		osBuildConfig.Status.LastKnownUserConfiguration = &userConfiguration
+		isNewOSBuildInstanceNeeded = true
+	}
+
+	if osBuildConfig.Annotations == nil {
```

Comment:
You can't return in this block because you are not updating the resource (if it changed).

Reply:
I think you didn't refresh - this was a commit that I pushed before I saw your comment (to patch once for annotation and user configuration)

Reply:
No, it's still here. If the configuration is different, but Annotations is nil you will return inside this block without updating the resource. I guess that in the UT Annotations is not nil.
BTW can it be nil? If not, than maybe this `if` is not needed